### PR TITLE
TEP-0118: Added Matrix.Include field in preview mode

### PR DESCRIPTION
**Pull request by EmmaMunley**
# Changes
[TEP-0118](https://github.com/tektoncd/community/blob/main/teps/0118-matrix-with-explicit-combinations-of-parameters.md) proposed adding include as a new field within matrix

In this change, we add the `matrix.Include` field to the matrix specification and documentation the API change - we will incrementally add further details as we iterate on the implementation of [TEP-0118](https://github.com/tektoncd/community/blob/main/teps/0118-matrix-with-explicit-combinations-of-parameters.md) to allow specify specific combinations or explicit combinations for Matrix Parameters. Validation and implementation logic will be added in subsequent PRs.

Note: This PR is to only add the new `Matrix.Include` field and update the documentation in "Preview Mode".  No functionality has been added at this time.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
API Change: Added Matrix.Include in preview mode (Not yet functional)
```

### CODE REVIEW

File: pkg/apis/pipeline/v1/pipeline_types.go [hunk 1]
```diff
@@ -212,6 +212,24 @@ type Matrix struct {
 	// The names of the `params` in the `Matrix` must match the names of the `params` in the underlying `Task` that they will be substituting.
 	// +listType=atomic
 	Params []Param `json:"params,omitempty"`
+
+	// Include is a list of MatrixInclude
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1/pipeline_types.go:216]: NIT: would have been helpful to describe it's purpose even in preview mode

File: pkg/apis/pipeline/v1beta1/pipeline_types.go [hunk 2]
```diff
@@ -170,6 +170,24 @@ type Matrix struct {
 	// The names of the `params` in the `Matrix` must match the names of the `params` in the underlying `Task` that they will be substituting.
 	// +listType=atomic
 	Params []Param `json:"params,omitempty"`
+
+	// Include is a list of MatrixInclude
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_types.go:174]: Similar to the comment ⬆️ , please add a brief explanation

**Review by pritidesai — APPROVED** [review]
thank you @EmmaMunley 👍 Exciting to have this feature implemented ... 🎉

File: docs/pipeline-api.md [hunk 3]
```diff
@@ -2,7 +2,7 @@
 ---
 title: Pipeline API
 linkTitle: Pipeline API
-weight: 404
+weight: 1000
```
Comment (jerop) [inline docs/pipeline-api.md:5]: why are you changing the weight? please revert this to 404

https://github.com/tektoncd/pipeline/pull/6200 fixes the script

File: pkg/apis/pipeline/v1/pipeline_types.go [hunk 4]
```diff
@@ -212,6 +212,24 @@ type Matrix struct {
...
+	// Note this field is in preview mode and not yet supported
```
Comment (jerop) [inline pkg/apis/pipeline/v1/pipeline_types.go:217]:
```suggestion
	// Include is a list of MatrixInclude which allows passing in specific combinations of Parameters into the Matrix.
	// Note that Include is in preview mode and not yet supported.
```

File: pkg/apis/pipeline/v1beta1/pipeline_types.go [hunk 5]
```diff
@@ -170,6 +170,24 @@ type Matrix struct {
...
+	// Note this field is in preview mode and not yet supported
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/pipeline_types.go:175]:
```suggestion
	// Include is a list of MatrixInclude which allows passing in specific combinations of Parameters into the Matrix.
	// Note that Include is in preview mode and not yet supported.
```

**Review by jerop — COMMENTED** [review]
Thank you @EmmaMunley! 

Requesting some docs changes, otherwise looks good to me

**Comment by pritidesai** [comment]
@EmmaMunley please fix the commits, its messed up with 14 commits 😞

File: pkg/apis/pipeline/v1/pipeline_types.go [hunk 6]
```diff
@@ -212,6 +212,24 @@ type Matrix struct {
...
+	Include []MatrixInclude `json:"include,omitempty"`
+}
+
+// MatrixInclude allows passing in a specific combinations of Parameters into the Matrix.
+// Note this struct is in preview mode and not yet supported
+type MatrixInclude struct {
+	// Name the specified combination
+	Name string `json:"name,omitempty"`
+
+	// Params takes only `Parameters` of type `"string"`
```
Comment (jerop) [inline pkg/apis/pipeline/v1/pipeline_types.go:229]: planning to add validation that the parameters are of type string in a subsequent PR?

**Review by jerop — COMMENTED** [review]
@EmmaMunley changes look good to me, but please update the commit message with more detail - https://github.com/tektoncd/community/blob/main/standards.md#commit-messages

File: pkg/apis/pipeline/v1/pipeline_types.go (hunk 6 above)
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipeline_types.go:229]: Updated commit message to call out that validation will be added in a subsequent PR.

**Review by jerop — COMMENTED** [review]
/lgtm

**Pull request merged**